client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java

```diff
@@ -165,8 +165,17 @@ protected static final void setApplicationProperties(final Message msg, final Ma
             sender.setQoS(qos);
             sender.openHandler(senderOpen -> {
                 if (senderOpen.succeeded()) {
-                    LOG.debug("sender open [{}]", sender.getRemoteTarget());
-                    result.complete(senderOpen.result());
+                    LOG.debug("sender open [{}] sendQueueFull [{}]", sender.getRemoteTarget(), sender.sendQueueFull());
+                    // wait on credits a little time, if not already given TODO: configurable waiting time
+                    if(sender.sendQueueFull()) {
+                        ctx.owner().setTimer(10, timerID -> {
+                            LOG.debug("waited 10ms on credits [{}], sendQueueFull [{}]", sender.getCredit(), sender.sendQueueFull());
```

Comment:
logging the credits and the *sendQueueFull* flag seems redundant, the flag is actually calculated as `sender.getCredit() <= 0` ...

Reply:
Have you played around with the timeout value? Do 10ms already do the trick in most cases?

Reply:
The sendQueueFull flag is implemented as link.getRemoteCredit() <= 0, which should include the queued messages. 

My idea was to start with something (which I played around locally) and then try this later in a more realistic environment. But now I think it might be better to directly implement it configurable. WDYT?

Reply:
> The sendQueueFull flag is implemented as link.getRemoteCredit() <= 0, which should include the queued messages.

you are right in that `ProtonSenderImpl.sendQueueFull()` invokes `LinkImpl.getRemoteCredit()`. FMPOV, this then means that the condition in the `if ()` statement is not appropriate because it also is `true` if the sender has no credit (yet) but has some space left in its outbound buffer (which should always be the case if the link has been freshly created as in this case).

Reply:
Here's proton-j `SenderLink.getRemoteCredit()`:
````
    public int getRemoteCredit()
    {
        // Credit is decremented as soon as advance is called on a send,
        // so we need only consider the credit count, not the queued count.
        return getCredit();
    }
````
Here's `getCredit()` JavaDoc:
````
     * Gets the credit balance for a link.
     *
     * Note that a sending link may still be used to send deliveries even if
     * link credit is/reaches zero, however those deliveries will end up being
     * {@link #getQueued() queued} by the link until enough credit is obtained
     * from the receiver to send them over the wire. In this case the balance
     * reported will go negative.
```` 
It seems that invoking `ProtonSender.sendQueueFull()` will return `false` if the underlying link has no credits yet or left because in this case `getCredit()` will return a value <= 0 (which is indeed the check done by `sendQueueFull()`.
So, the `if ()` statement should be ok and it seems that logging `getCredit()` and `sendQueueFull()` will yield synonymous results ...

Reply:
I think this is not implemented or described correct in Proton. This is the reason, why I wanted to log this both values, to have this if the implementation might change. In fact the getRemoteCredit() from the SenderImpl calls getCredit() and does not include any queued values. They also commented it this way. So implementation and API description are not fully matched, but they seem to do this intentionally.
But maybe we should change then our code to only check on getCredit(), since this is what we really want to check. But then we also should change the other occurrences  - e.g. in AbstractSender.send()?

Reply:
Oh - we had race conditions in commenting. :)

Reply:
I do not think that the API doc for proton-j is inconsistent. In fact, here's `Link.getCredit()`:
````
     * Gets the credit balance for a link.
     *
     * Note that a sending link may still be used to send deliveries even if
     * link credit is/reaches zero, however those deliveries will end up being
     * {@link #getQueued() queued} by the link until enough credit is obtained
     * from the receiver to send them over the wire. In this case the balance
     * reported will go negative.
     *
     * @return the credit balance for the link
 ````
It does not say anything about how the credit value is actually calculated (i.e. whether it already accounts for the queued messages or not). It only states that the returned value will be negative if a message being sent will end up in the outbound buffer (queue).
`SenderImpl` seems to count down the credit as soon as `send` is invoked, no matter if the message actually gets sent over the wire or put to the buffer. But this is an implementation detail only. In this case it means that using `getCredit()` as the implementation of `getRemoteCredit()` yields the same result (which is consistent with the doc for both operations).
Thus, I think that we can safely keep using `ProtonSender.sendQueueFull()` ...

Reply:
I referred only to getRemoteCredit() and that the API description says that it considers getCredit() AND getQueued() but the implementation only uses getCredit() - but with a correct result as commented in the implementation (because "decremented in advance"). 

For me this is strange, because the semantics of getCredit() is then the same as getRemoteCredit(), but differently described.

But anyway - we can stay with sendQueueFull() and trust the API.